contributors/design-proposals/code-generator/kubegen.md

```diff
@@ -0,0 +1,269 @@
+
+```sh
+kubegen --license-file "boilerplate.go.txt"
+```
+
+- run code generators against discovered API type definitions
+- prepend license to generated files
+  - use the contents from the boilerplate.go.txt file instead of LICENSE
+  - if boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments
+  - exit non-zero if boilerplate.go.txt is missing
+  
+**Note**: This passes the `--go-header-file` to each of the code generators
+
+#### Specifying a license type
+
+The license can be overridden from the LICENSE file using the `--license` flag
+
+```sh
+kubegen --license "Apache 2.0"
+```
+
+- run code generators against discovered API type definitions
+- prepend license to generated files
+  - use an *Apache 2.0* license
+
+#### Specifying external API versions
+
+Generating code only for specific API group versions can be performed by providing the API
+group versions as positional arguments.  When manually specifying the API group versions,
+neither internal nor external API types are automatically discovered.
+
+```sh
+kubegen apps/v1 apps/v1beta1 extensions/v1beta1
+```
+
+- run all code generators against the apps/v1 apps/v1beta1 extensions/v1beta1 API groups versions
+  - only run against external types, do not run against internal types
+- prepend license to generated files
+
+#### Specifying internal API groups
+
+Generating code only for specific API groups can be performed by providing the API
+groups as positional arguments.  When manually specifying the API groups,
+neither internal nor external API types are automatically discovered.
+
+```sh
+kubegen apps extensions
+```
+
+- run all code generators against the apps extensions API groups
+  - only run against internal types, do not run against external types
+- prepend license to generated files
+
+#### Specifying both internal API groups and external API versions
+
+To generate code for both manually specified internal and external types
+provide both as positional arguments.
+
+```sh
+kubegen apps apps/v1 apps/v1beta1 extensions extensions/v1beta1
+```
+
+- run all code generators against the provided API groups and versions
+  - generate internal types for apps and extensions groups
+  - generate external types for apps/v1 apps/v1beta1 extensions/v1beta1 versions
+- prepend license to generated files
+
+#### Running in dry-run
+
+To verify the output without writing actual files use the `--dry-run` flag
+
+```sh
+kubegen --dry-run
+```
+
+**Note**: This passes the `--verify-only` flag to each of the code generators
+
+  
+#### Specifying where to search for APIs
+
+The `--apis-dir` defaults to `pkg/apis` and looks for API groups in that directory.
+
+Looking for APIs outside the default location can be configured using
+the `--apis-dir` flag.  This flag may be provided multiple times to search multiple
+directories.
+
+```sh
+kubegen --apis-dir notpkg/apis --apis-dir pkg/notapis
+```
+
+- run all code generators against discovered APIs
```

Comment:
discovered api mean `types.go` exists?

Reply:
No, I don't think it should since folks may want to break up a monolithic types.go file.  I was thinking purely based off directory structure, and then the presence of +genclient